From a6d1a6d252c7afe801816defabf4369d8be268ff Mon Sep 17 00:00:00 2001
From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Date: Thu, 13 Feb 2025 06:46:30 +0900
Subject: [PATCH] ntdll: Register module dependency for export forwarders
 regardless of whether the dependency is already loaded.

Prepare for generalization to imports from DLLs.

Calling add_module_dependency() multiple times for the same dependency
edge no longer bloats memory usage.
---
 dlls/ntdll/loader.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index c0cd5d41a05..cce46090754 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -979,17 +979,18 @@ static FARPROC find_forwarded_export( HMODULE module, const char *forward, LPCWS
             forward, debugstr_w(get_modref(module)->ldr.FullDllName.Buffer),
             debugstr_w(get_modref(module)->ldr.BaseDllName.Buffer) );
     }
-    else if (wm_owned_ref)
+    else if (wm->ldr.DdagNode != node_ntdll && wm->ldr.DdagNode != node_kernel32)
     {
         if (!(wm->ldr.Flags & LDR_DONT_RESOLVE_REFS))
         {
             if (!imports_fixup_done && importer)
             {
                 /* Prepare for the callee stealing the reference */
-                wm_owned_ref = FALSE;
+                if (wm_owned_ref) wm_owned_ref = FALSE;
+                else if (wm->ldr.LoadCount != -1) wm->ldr.LoadCount++;
                 add_module_dependency( importer->ldr.DdagNode, wm->ldr.DdagNode );
             }
-            else if (process_attach( wm->ldr.DdagNode, NULL ) != STATUS_SUCCESS)
+            else if (wm_owned_ref && process_attach( wm->ldr.DdagNode, NULL ) != STATUS_SUCCESS)
             {
                 proc = NULL;
             }
-- 
GitLab

